File: python/runners/speech-to-text/speech_to_text.py
"""Transcribe speech in an audio (or video) file with a locally pulled Hugging Face transformers model.

Compiled into a standalone binary with PyInstaller, so it runs without a Python
installation. Designed to be driven by the PHP wrapper:

  * stdout: one JSON object on success, e.g. {"text": "he hoped there would be stew for dinner"},
            plus "segments": [{"start": 0.0, "end": 2.5, "text": "..."}, ...] with --timestamps
  * stderr: progress and error messages for humans
  * exit code: 0 success, 1 transcription failed, 2 invalid arguments

Both kinds of speech recognition models are supported:
  * sequence-to-sequence models, e.g. Whisper; multilingual ones can be told the spoken language
    and can translate the speech into English
  * CTC models, e.g. Wav2Vec2, HuBERT; usually for a single language

Audio is decoded with PyAV, which bundles FFmpeg's libraries: WAV, MP3, M4A/AAC, FLAC, OGG/Opus
and the audio track of video files work without FFmpeg being installed.
"""

import argparse
import json
import multiprocessing
import os
import sys
import traceback

EXIT_OK = 0
EXIT_FAILED = 1

# Whisper reads 30-second windows; longer audio is transcribed window by window.
CHUNK_SECONDS = 30


class UserError(Exception):
    """A problem the user can fix, reported without a stack trace."""


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="speech-to-text",
        description="Transcribe speech in an audio or video file with a locally pulled transformers model.",
    )
    parser.add_argument("--model", required=True, help="Directory of a pulled transformers model (contains config.json)")
    parser.add_argument("--audio", required=True, help="Path of the audio or video file, e.g. interview.mp3")
    parser.add_argument("--language", help="Spoken language for multilingual models such as Whisper, e.g. en, french (default: detected)")
    parser.add_argument("--translate", action="store_true", help="Translate the speech into English (Whisper-style models)")
    parser.add_argument("--timestamps", action="store_true", help="Also report when each segment is spoken")
    parser.add_argument("--device", help="Torch device, e.g. cpu, cuda, mps (default: the best available)")
    return parser.parse_args(argv)


def log(message: str) -> None:
    print(message, file=sys.stderr, flush=True)


def load_audio(path: str, sampling_rate: int):
    """Decodes the first audio track of any FFmpeg-readable file into mono float32 samples at sampling_rate."""
    import av
    import numpy as np

    try:
        container = av.open(path)
    except av.error.FileNotFoundError as e:
        raise UserError(f"{path} does not exist") from e
    except av.error.InvalidDataError as e:
        raise UserError(f"{path} is not an audio or video file in a format FFmpeg can read") from e

    with container:
        stream = next((s for s in container.streams if s.type == "audio"), None)
        if stream is None:
            raise UserError(f"{path} has no audio track")

        resampler = av.AudioResampler(format="flt", layout="mono", rate=sampling_rate)
        chunks = []
        for frame in container.decode(stream):
            for resampled in resampler.resample(frame):
                chunks.append(resampled.to_ndarray().reshape(-1))
        for resampled in resampler.resample(None):
            chunks.append(resampled.to_ndarray().reshape(-1))

    if not chunks:
        raise UserError(f"{path} contains no audio samples")

    return np.concatenate(chunks).astype(np.float32)


def main(argv: list[str]) -> int:
    args = parse_args(argv)

    # Models are loaded from disk only; never reach out to the Hub.
    os.environ["HF_HUB_OFFLINE"] = "1"

    # Imported after argument parsing so --help and usage errors stay fast.
    import torch
    from transformers import pipeline
    from transformers.utils import logging as transformers_logging

    # Progress bars redraw with carriage returns, which garbles pipes and log files.
    if not sys.stderr.isatty():
        transformers_logging.disable_progress_bar()

    try:
        device = args.device or best_device(torch)
        dtype = torch.float16 if device.startswith("cuda") else torch.float32

        log(f"[speech-to-text] Loading {args.model} on {device}...")
        asr = pipeline("automatic-speech-recognition", model=args.model, device=device, dtype=dtype)

        sampling_rate = asr.feature_extractor.sampling_rate
        audio = load_audio(args.audio, sampling_rate)
        seconds = len(audio) / sampling_rate

        options = {}
        if asr.model.config.is_encoder_decoder:
            generate_kwargs = {}
            if args.language or args.translate:
                if not getattr(asr.model.generation_config, "lang_to_id", None):
                    raise UserError("this model cannot be told a language or translate; it only transcribes the language it was trained on")
                if args.language:
                    generate_kwargs["language"] = args.language
                if args.translate:
                    generate_kwargs["task"] = "translate"
            options["generate_kwargs"] = generate_kwargs
            # Whisper transcribes audio longer than one window sequentially, which needs segment timestamps.
            if args.timestamps or seconds > CHUNK_SECONDS:
                options["return_timestamps"] = True
        else:
            if args.language or args.translate:
                raise UserError("this model cannot be told a language or translate; it only transcribes the language it was trained on")
            if seconds > CHUNK_SECONDS:
                options["chunk_length_s"] = CHUNK_SECONDS
            if args.timestamps:
                options["return_timestamps"] = "word"

        log(f"[speech-to-text] Transcribing {seconds:.1f} seconds of audio...")
        output = asr({"raw": audio, "sampling_rate": sampling_rate}, **options)
    except UserError as e:
        log(f"[speech-to-text] Error: {e}")
        return EXIT_FAILED
    except Exception as e:
        traceback.print_exc()
        log(f"[speech-to-text] Error: {type(e).__name__}: {e}")
        return EXIT_FAILED

    result = {"text": output["text"].strip()}
    if args.timestamps:
        result["segments"] = [
            {"start": chunk["timestamp"][0], "end": chunk["timestamp"][1], "text": chunk["text"].strip()}
            for chunk in output.get("chunks", [])
        ]

    print(json.dumps(result), flush=True)
    log("[speech-to-text] Done.")
    return EXIT_OK


def best_device(torch) -> str:
    if torch.cuda.is_available():
        return "cuda"
    if torch.backends.mps.is_available():
        return "mps"
    return "cpu"


if __name__ == "__main__":
    # torch starts multiprocessing helpers by re-running sys.executable, which in the frozen binary is this
    # script; freeze_support() lets those helper invocations run instead of reaching the argument parser.
    multiprocessing.freeze_support()
    sys.exit(main(sys.argv[1:]))

File: src/Runner/SpeechToText.php
<?php

declare(strict_types=1);

namespace PhpLovesAi\Runner;

use PhpLovesAi\Binary\Tool;
use PhpLovesAi\Exception\AudioNotFoundException;
use PhpLovesAi\Exception\BinaryNotInstalledException;
use PhpLovesAi\Exception\HomeDirectoryNotFoundException;
use PhpLovesAi\Exception\ModelNotFoundException;
use PhpLovesAi\Exception\RunFailedException;
use PhpLovesAi\Exception\UnsupportedModelException;
use PhpLovesAi\Filesystem\Path;

/**
 * Transcribes speech in audio or video files with transformers models pulled into the project, through the runner
 * installed by `vendor/bin/loves-ai setup speech-to-text`.
 *
 *     $transcript = (new SpeechToText())->transcribe('openai/whisper-tiny', __DIR__ . '/interview.mp3');
 *
 * Works with sequence-to-sequence models (e.g. Whisper, which can be told the spoken language and translate into
 * English) and CTC models (e.g. Wav2Vec2), from the Hugging Face task "automatic-speech-recognition". Audio of any
 * length and common formats (WAV, MP3, M4A, FLAC, OGG, the audio track of videos) is decoded by the runner itself.
 */
final class SpeechToText extends BinaryRunner
{
    /** A model the runner can load; used in error messages and hints. */
    public const EXAMPLE_MODEL = 'openai/whisper-tiny';

    /** Hugging Face models this runner can use. */
    public const COMPATIBLE_MODELS_URL = 'https://huggingface.co/models?pipeline_tag=automatic-speech-recognition&library=transformers';

    public static function tool(): Tool
    {
        return Tool::SpeechToText;
    }

    /**
     * @param string                        $model     Hugging Face model id, e.g. "openai/whisper-tiny"
     * @param string                        $audioPath audio or video file, e.g. "interview.mp3"; a leading "~" is expanded
     * @param string|null                   $language  spoken language for multilingual models such as Whisper, e.g.
     *                                                 "en" or "french"; null detects it
     * @param bool                          $translate translate the speech into English (Whisper-style models)
     * @param string|null                   $device    torch device such as "cpu", "cuda" or "mps"; null picks the best
     * @param (callable(string): void)|null $onOutput  receives the runner's progress output as it arrives
     *
     * @return string the transcript
     *
     * @throws AudioNotFoundException
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException when the model is not a transformers speech recognition model, or a language
     *                                   or translation is asked of a model that only knows one language
     * @throws HomeDirectoryNotFoundException
     * @throws RunFailedException e.g. when the file has no audio
     */
    public function transcribe(
        string $model,
        string $audioPath,
        ?string $language = null,
        bool $translate = false,
        ?string $device = null,
        ?callable $onOutput = null,
    ): string {
        return $this->transcription($model, $audioPath, $language, $translate, false, $device, $onOutput)['text'];
    }

    /**
     * Like transcribe(), split into timed segments: sentences or phrases for Whisper-style models, words for CTC models.
     *
     * @return list<array{start: float, end: float|null, text: string}> seconds from the start of the audio; the last
     *                                                                   segment may have no end
     *
     * @throws AudioNotFoundException
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException
     * @throws HomeDirectoryNotFoundException
     * @throws RunFailedException
     */
    public function transcribeWithTimestamps(
        string $model,
        string $audioPath,
        ?string $language = null,
        bool $translate = false,
        ?string $device = null,
        ?callable $onOutput = null,
    ): array {
        return $this->transcription($model, $audioPath, $language, $translate, true, $device, $onOutput)['segments'];
    }

    /**
     * Runs the checks transcribe() performs before starting the binary, including whether the model can be told a
     * language or translate, without transcribing anything.
     *
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException
     */
    public function ensureCanTranscribe(string $model, ?string $language = null, bool $translate = false): void
    {
        $this->ensureCanRun($model);

        // Multilingual models such as Whisper list the languages they can be told in generation_config.json.
        $generationConfig = json_decode((string) @file_get_contents($this->modelPath($model) . '/generation_config.json'), true);
        if (($language !== null || $translate) && !isset($generationConfig['lang_to_id'])) {
            throw new UnsupportedModelException($model, $this->modelPath($model), sprintf(
                '%s cannot be told the spoken language or translate: it only transcribes the language it was trained on. '
                . 'Leave out the language and translation, or use a multilingual model such as %s.',
                $model,
                self::EXAMPLE_MODEL,
            ));
        }
    }

    /**
     * The runner loads models with transformers; whisper.cpp and faster-whisper conversions of the same models are
     * common on Hugging Face but need other software.
     */
    protected function ensureModelIsSupported(string $model, string $modelPath): void
    {
        $problem = match (true) {
            !TransformersModel::has($modelPath, 'config.json') && TransformersModel::has($modelPath, 'ggml*.bin')
                => 'it is in whisper.cpp (GGML) format, which transformers cannot load.',
            TransformersModel::has($modelPath, 'model.bin') && !TransformersModel::has($modelPath, '*.safetensors')
                && !TransformersModel::has($modelPath, 'pytorch_model*.bin')
                => 'it is in faster-whisper (CTranslate2) format, which transformers cannot load.',
            default => TransformersModel::problem($modelPath) ?? self::audioProblem($modelPath),
        };

        if ($problem === null) {
            return;
        }

        throw new UnsupportedModelException($model, $modelPath, sprintf(
            '%s cannot transcribe speech: %s Use a transformers speech recognition model instead, e.g. %s (browse: %s).',
            $model,
            $problem,
            self::EXAMPLE_MODEL,
            self::COMPATIBLE_MODELS_URL,
        ));
    }

    /**
     * Speech models read audio through a feature extractor, configured in preprocessor_config.json (or, in some older
     * repositories, feature_extractor_config.json) with at least the audio's sampling rate.
     */
    private static function audioProblem(string $modelPath): ?string
    {
        $configs = [];
        foreach (['preprocessor_config.json', 'feature_extractor_config.json'] as $file) {
            $config = json_decode((string) @file_get_contents("{$modelPath}/{$file}"), true);
            if (is_array($config)) {
                $configs[] = $config;
            }
        }

        foreach ($configs as $config) {
            if (isset($config['feature_extractor_type']) || isset($config['sampling_rate'])) {
                return null;
            }
        }

        foreach ($configs as $config) {
            if (isset($config['image_processor_type'])) {
                return 'it reads images, not audio; use it with image-to-text.';
            }
        }

        return 'it cannot hear audio (it has no audio feature extractor in preprocessor_config.json). If it is a text model, use it with text-to-text.';
    }

    /**
     * @return array{text: string, segments: list<array{start: float, end: float|null, text: string}>}
     *
     * @throws AudioNotFoundException
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException
     * @throws HomeDirectoryNotFoundException
     * @throws RunFailedException
     */
    private function transcription(
        string $model,
        string $audioPath,
        ?string $language,
        bool $translate,
        bool $timestamps,
        ?string $device,
        ?callable $onOutput,
    ): array {
        $audioPath = Path::expandHome($audioPath);
        if (!is_file($audioPath) || !is_readable($audioPath)) {
            throw new AudioNotFoundException($audioPath);
        }

        $this->ensureCanTranscribe($model, $language, $translate);

        $result = $this->run($model, [
            '--audio' => $audioPath,
            '--language' => $language,
            '--translate' => $translate,
            '--timestamps' => $timestamps,
            '--device' => $device,
        ], $onOutput);

        if (!is_string($result['text'] ?? null)) {
            throw new RunFailedException(0, 'The runner did not report the transcript.');
        }

        $segments = [];
        foreach ($timestamps && is_array($result['segments'] ?? null) ? $result['segments'] : [] as $segment) {
            if (is_array($segment) && is_numeric($segment['start'] ?? null) && is_string($segment['text'] ?? null)) {
                $segments[] = [
                    'start' => (float) $segment['start'],
                    'end' => is_numeric($segment['end'] ?? null) ? (float) $segment['end'] : null,
                    'text' => $segment['text'],
                ];
            }
        }

        return ['text' => $result['text'], 'segments' => $segments];
    }
}
